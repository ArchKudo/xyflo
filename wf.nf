#!/usr/bin/env nextflow
nextflow.enable.dsl = 2

// Create a commandline parameter for NCBI's BioProject identifier
// with default value as PRJNA1005421
params.bioProjectID = 'PRJNA1005421'

// Create a commandline parameter for reference sequences fasta file
// with default value as xylo.fna
params.referenceSequences = 'data/xylo.fna'

process fetchRunAccesionsForBioProject {
    output:
        path 'runAccessions.txt'

    script:
    """
    wf.sh fetchRunAccesionsForBioProject "${params.bioProjectID}" > "runAccessions.txt"
    """
}

process downloadSRAForRunAccession {
    input:
        val runAccession
    output:
        tuple val("${runAccession}"), path("data/${runAccession}/${runAccession}.sra")
    script:
    """
    wf.sh downloadSRAForRunAccession "$runAccession"
    """
}

process extractFASTQFromSRAFile {
    input:
        tuple val(runAccession), val(sraFilePath)
    output:
        tuple val("${runAccession}"),
        path("pairs/${runAccession}_1.fastq"),
        path("pairs/${runAccession}_2.fastq")

    script:
    """
    # TODO: Don't hardcode mem and threads requirement
    wf.sh extractFASTQFromSRAFile "$sraFilePath" "5" "12"
    """
}

process buildBowtieIndexFromFASTA {
    input:
        path referenceSequencesFASTA
    output:
        tuple path('db'), val("$name")
    script:
        name = "$referenceSequencesFASTA.fileName.name"
        """
        mkdir -p db/
        wf.sh buildBowtieIndexFromFASTA "$referenceSequencesFASTA" "db/$name" --threads 12
        """
}

process alignRunWithBowtie {
    input:
        // TODO: Use https://www.nextflow.io/docs/latest/process.html#dynamic-input-file-names
        tuple val(runAccession), path(pair_1), path(pair_2), path(referenceSequences), val(referenceSequencesName)
    output:
        path("aln/${runAccession}.sam")

    script:
    """
    mkdir -p aln/
    wf.sh alignRunWithBowtie "$runAccession" "$pair_1" "$pair_2" "$referenceSequences/$referenceSequencesName" "12"
    """
}

process mergeSAMFiles {
    input:
        path(alignedSAMFile)
    output:
        path('merged.sam')
    script:
    args = alignedSAMFile.collect { "-I $it" }.join(' ')
    """
    picard MergeSamFiles $args -O "merged.sam"
    """
}

process convertSAMToBAM {
    input:
        path(mergedSAMFile)
    output:
        path('merged.bam')
    script:
    """
    samtools view -@ 16 --verbosity 100 -b "$mergedSAMFile" -o "merged.bam"
    """
}

process sortBAMFile {
    input:
        path(mergedBAMFile)
    output:
        path('sorted.bam')
    script:
    """
    samtools sort -m 512M --threads 16 --verbosity 100 -o sorted.bam "$mergedBAMFile"
    """
}

process deduplicateBAMFile {
    input:
        path(sortedBAMFile)
    output:
        path('dedup.sam')
    script:
    """
    picard.jar MarkDuplicates INPUT="$sortedBAMFile" OUTPUT=dedup.bam METRICS_FILE=dedup.metrics
    """
}

process createBAMIndexFile {
    input:
        path(bamFile)
    output:
        tuple path(bamFile), path("${bamFile}.bai")
    script:
    """
    samtools index "$bamFile"
    """
}

process pileUpAndVariantCallBAM {
    input:
        tuple path(referenceSequence), path(targetBAMFile)
    output:
        path('calls.vcf')
    script:
    """
    bcftools mpileup -Ou --threads 12 -f "$referenceSequence" "$targetBAMFile" | \
    bcftools call --threads 12 -mv -Oz -o calls.vcf
    """
}

process recoverHaplotypes {
    errorStrategy 'ignore'
    input:
        tuple path(bamFile), path(bamFileIndex), path(referenceSequence)
    output:
        path('out.fasta')
    script:
    """
    wf.sh recoverHaplotypes "$bamFile" "$referenceSequence" "$projectDir/bin/snpper.py"
    """
}

workflow {
    seqChan = Channel.fromPath(params.referenceSequences)
    index = buildBowtieIndexFromFASTA(seqChan)
    wf = fetchRunAccesionsForBioProject
        | map { it.readLines() }
        | flatten
        | take(3)
        | downloadSRAForRunAccession
        | extractFASTQFromSRAFile
        | combine(index)
        | alignRunWithBowtie
        | flatten
        | toList
        | mergeSAMFiles
        | convertSAMToBAM
        | createBAMIndexFile
        | combine(seqChan)
        // | flatten
        | recoverHaplotypes
        | view

    // Use snpper instead
    // variant = pileUpAndVariantCallBAM(Channel.fromPath(params.referenceSequences).combine(align.first()))
    // variant | view
}
