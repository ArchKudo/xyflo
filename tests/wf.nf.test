nextflow_process {

    name "Test Process fetchRunAccesionsForBioProject"
    script "wf.nf"
    process "fetchRunAccesionsForBioProject"

    test("Should run without failures") {

        when {
            params {
                // define parameters here. Example:
                // outdir = "tests/results"
            }
            process {
                """
                // define inputs of the process here. Example:
                input[0] = "PRJNA732531"
                """
            }
        }

        then {
            assert process.success
            assert snapshot(process.out).match()
        }

    }

}
