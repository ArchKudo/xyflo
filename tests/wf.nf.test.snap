{
    "Should run without failures": {
        "content": [
            {
                "0": [
                    "Run\nSRR14710930\nSRR14710929\nSRR14710918\nSRR14710907\nSRR14710896\nSRR14710887\nSRR14710886\nSRR14710885\nSRR14710884\nSRR14710883\nSRR14710928\nSRR14710927\nSRR14710926\nSRR14710925\nSRR14710924\nSRR14710923\nSRR14710922\nSRR14710921\nSRR14710920\nSRR14710919\nSRR14710917\nSRR14710916\nSRR14710915\nSRR14710914\nSRR14710913\nSRR14710912\nSRR14710911\nSRR14710910\nSRR14710909\nSRR14710908\nSRR14710906\nSRR14710905\nSRR14710904\nSRR14710903\nSRR14710902\nSRR14710901\nSRR14710900\nSRR14710899\nSRR14710898\nSRR14710897\nSRR14710895\nSRR14710894\nSRR14710893\nSRR14710892\nSRR14710891\nSRR14710890\nSRR14710889\nSRR14710888\n"
                ]
            }
        ],
        "timestamp": "2023-10-01T12:33:57+0000"
    }
}